### ISSUE TITLE
Map background layers

### ISSUE THREAD

**Issue by asaunier** [c2corg/v6_ui#67]
In v5, maps are based upon Gmaps relief (by default), OSM, Swisstopo (for Switzerland) and IGN (for France).

In v6, we use OpenLayers3. Which does not come with GMaps layers, because of some conflicts with GMaps terms of use.

If we don't use Gmaps, we need to find an alternate default background layer. Either free such as OSM (but their are limitations concerning the number of tiles that can be loaded) or paying ones (the fee depending on the number of tiles that can be loaded).

Free (?) layers:
* OSM
* Esri maps: http://arcg.is/1TRMeJJ / http://blogs.esri.com/esri/arcgis/2011/02/22/use-arcgis-online-basemaps-for-free-in-your-gis-projects-and-web-applications/
* Bing maps (not sure "relief" maps are available)
* ?

Paying options:
* https://www.mapbox.com/blog/mapbox-outdoors/
* ??

Using an extension of OL3 for using Gmaps layers:
* http://www.mapgears.com/fr/blogue/archive/2015-10-23-ol3-google-maps

Please note that when switching among background layers, we might have to change the map projection (for instance when using Swisstopo WMTS).

**Comment by tsauerwein** [comment]
ESRI base maps seem to be free for non-commercial sites:
https://github.com/leaflet-extras/leaflet-providers/issues/22#issuecomment-28003372

> Development and Testing Plans are generally intended for developing and testing your applications for ArcGIS Online. **However, you can also make your applications powered by your Development and Testing Plan available to the public, so long as you don’t monetize your application in any way**, and it's not used for the benefit of a business or government agency.
https://developers.arcgis.com/terms/

**Comment by fjacon** [comment]
@asaunier  : Should we ask for confirmation ? What is the position of the client ?

**Comment by asaunier** [comment]
I don't really know. I can ask.

**Comment by asaunier** [comment]
Not much input from the client...

Any way we need the same IGN and Swisstopo baselayers than in v5 in addition to the OSM layer already available in v6:
* we need a baselaer selector (available in ngeo)
* we need to support reprojection (Swisstopo baselayers do not use EPSG:3857 as OSM, IGN, GMaps...)

Having Google Maps would be much appreciated but would require ol3-google-maps:
https://github.com/mapgears/ol3-google-maps/
Do you think it would cause problems to use it instead of a standard OL3?

**Comment by tsauerwein** [comment]
I gave ol3-google-maps a try to better understand how it works. It allows to use Google Maps with OpenLayers without having to get in touch with the Google Maps API. You basically add a new layer `olgm.layer.Google` to your map and  ol3-google-maps takes care of the rest.

So, how does it work? If you add a GMaps layer, it creates a new GMaps map instance in the background and adds the original ol map as control to the GMaps map. All interaction is done with the ol map. If you drag/zoom the map, the GMaps map is dragged/zoomed in the background. To make this synchronization more smooth, vector layers are replicated in the GMaps map. If you add a vector layer to the ol map, the vector layer is made transparent and a GMaps vector layer is created in the GMaps map. For each feature, a corresponding GMaps feature is created. The ol styles are converted to GMaps styles.

Here are two examples: [examples/concept.html](http://dev5.mapgears.com/ol3-google-maps/examples/concept.html) or [draw/modify](http://bl.ocks.org/tsauerwein/raw/ff5deb1d50b9de2b5b5661f7ea95ae25/index.html) (more [here](http://dev5.mapgears.com/ol3-google-maps/examples/))

Some limitations are listed [here](https://github.com/mapgears/ol3-google-maps/blob/master/LIMITATIONS.md), relevant for us are the followings:

- No kinetic pan. Panning feels like it did 10 years ago.
- No support for rotated maps.
- No support for ol style functions.
- We are limited to the styles that are both supported by ol and GMaps. We couldn't take advantage of the nice styling features of ol.
- Editing features feels slow and you see a lag (for example try to drag a point with a label in the concept example, or modifying a line in the other example).
- The GMaps layers (GMaps tiles + features) are always in the background. If you add another tile layer (e.g. the avalanche WMS layer) the features would be behind the WMS layer.
- The support for GMaps in ol2 was always fragile. A change in the GMaps API could break the support for ol. This is still the case.
- It is unclear how well ol3-google-maps deals with many features. Because features are added to an ol vector layer and also to a GMaps vector layer, I doubt that it works well for a lot of features (which will be the case for us).

In my opinion, ol3-google-maps is great for simple almost static maps. But in our case, only to have background tiles from Google Maps is not enough motivation to accept the limitations listed above.

We have tiles from IGN and Swisstopo for France and Switzerland, Bing has good satellite imagery, ESRI has a nice terrain layer and then there is still OSM. IMHO we don't need tiles from Google Maps.

**Comment by asaunier** [comment]
Thanks for the investigations and your opinion @tsauerwein !
It seems then using Google Maps with OL3 (or ol3-google-maps) should be avoided.

**Comment by asaunier** [comment]
The ESRI baselayer visible at http://www.arcgis.com/home/webmap/viewer.html?webmap=a72b0766aea04b48bf7a0e8c27ccc007&extent=5.9387,45.5951,5.9999,45.6205 seems to be "World_Topo_Map"

See
* http://services.arcgisonline.com/arcgis/rest/services/World_Topo_Map/MapServer
* https://www.arcgis.com/home/item.html?id=30e5fe3149c34df1ba922e6f5bbf808f

Terms of Use: http://downloads2.esri.com/ArcGISOnline/docs/tou_summary.pdf
> Use ArcGIS Online maps and map layers, imagery, and geometry services **in conjunction with Esri software or an ArcGIS Online subscription** for projects and applications that are both internal and external use
As long as you do not exceed the transaction cap if the cap applies to you.
 Government, Education, and NGO users: Full unlimited use
 Private organizations, Business Partners, and Developers: 50,000,000 transactions per year. For maps, a
transaction is defined as eight tiles. For imagery and geometry services, a transaction is defined as one request.
For application scenarios exceeding 50 million transactions per year, additional fees apply

In short

> Use with Esri software, and comply with its terms of use. If you do not have Esri software, you must purchase an ArcGIS Online subscription

**Comment by stef74** [comment]
Ok Go pour OSM as default layer.

Association have to work on map subject.

2016-06-16 14:31 GMT+02:00 Alex Saunier <notifications@github.com>:

**Comment by asaunier** [comment]
Some additional info from other c2c SA projects:

baselayer "World_Topo_Map" has been successfully used in OL3 as WMTS tiles using https://server.arcgisonline.com/arcgis/rest/services/World_Topo_Map/MapServer/WMTS/1.0.0/WMTSCapabilities.xml

Concerning the usage restrictions:
https://doc.arcgis.com/fr/arcgis-online/reference/transaction-limits.htm
> Les organisations sont limitées à un total de 50 millions de transactions de recherche de fonds de carte et de géocodage ArcGIS sur une période de 12 mois.

(1 transaction = 8 tiles)

**Comment by asaunier** [comment]
Suggested by @LaurentBonnaud at https://github.com/c2corg/v6_ui/issues/438
> C2Cv6 uses OSM maps that are fine, but lack contour lines (courbes de niveau), which is problem for a mountaineering site. Would it be possible to use those maps instead (they are based on OSM data, but include contour lines)?
http://www.4umaps.eu/online-outdoor-hike-bicycle-map.aspx

**Comment by asaunier** [comment]
Swisstopo tiles are also available with the standard Mercator projection (EPSG: 3857), used for OSM, Esri, IGN etc. This avoids having to reproject all features displayed on the map when switching to a Swisstopo background. The standard Swisstopo projection is EPSG:21781.

See https://api3.geo.admin.ch/services/sdiservices.html#other-projections

I have used this version of the tiles for the baselayer selector created by #439.

Please note however that:
> Reprojected tiles are generated on-the-fly with MapProxy. If you plan to heavily use this service, please inform us in advance.

which means that the Swisstopo 3857 tiles might be displayed a little slower than the standard 21781 ones. As far as I can it is still OK (many of the reprojected tiles are probably already cached).
What about the "heavy use of the service"? As in v5, PR #439 makes the swisstopo tiles available only to auth users.

**Comment by LaurentBonnaud** [comment]
Here is a list to consider for other countries:

https://en.wikipedia.org/wiki/List_of_online_map_services

**Comment by stef74** [comment]
@shortoland Could you test and confirm we can close this issue. Don't hesitate to ask question if is not clear. For me it's ok I only open a issue becasue defaut zoom are too important.

**Comment by asaunier** [comment]
> defaut zoom are too important.

@stef74 What do you mean? What's the connection with the map background layers?

**Comment by stef74** [comment]
I have create a specific issue si not directly relying

2016-07-20 8:45 GMT+02:00 Alex Saunier <notifications@github.com>:

**Comment by shortoland** [comment]
@stef74 According to me OSM is now installed and is working well, so I close the ticket.
